### ISSUE TITLE
Upstreaming: Use greatestLowerBound in CFAbstractValue

### ISSUE THREAD

**Issue by cpovirk** [jspecify/checker-framework#5]
https://github.com/jspecify/checker-framework/commit/c5f5118df3bf7a3de2dcb9608db12c4481f70c69

The upstream code is essentially implementing GLB, anyway, only by using `isSubtype`.

(It turned out to be nice that I had had to migrate from `MultiGraphQualifierHierarchy` to `ElementQualifierHierarchy`: That had forced me to implement `greatestLowerBound` myself instead of inheriting an implementation that would likely have had the same problem as the `CFAbstractValue` code.)

But! `CFAbstractValue` has a concept of a "backup." Still, I'm hoping that we could preserve that by upstreaming a slightly less aggressive version of the above commit: We would still verify that `a <: b` or vice versa. But then, instead of inserting whichever one we detect to be the subtype first, we would still insert `greatestLowerBound(a, b)`. That would be equivalent to the existing code for reasonable type systems, but it would still work for our unreasonable type system :)

**Comment by cpovirk** [comment]
(But presumably that has some performance cost. So maybe upstreaming even a modified patch is not a slam dunk.)

**Comment by cpovirk** [comment]
(I may have overstated that: The usual implementation of `greatestLowerBound` is primarily just some `Map` lookups -- not free, but probably not the most expensive part of dataflow.)

**Comment by cpovirk** [comment]
I experimented with a possible alternative to this: I could ensure that our `CFAbstractValue` instances always use the `CFAnalysis` (and therefore the `NullSpecAnnotatedTypeFactory`, etc.) from the least convenient world.

(I had some problems with getting everything initialized in the right order. The approach I settled on was to override `createFlowAnalysis` to return a `CFAnalysis` subclass that overrides `createAbstractValue` to call `defaultCreateAbstractValue` but to pass `withLeastConvenientWorld().analysis` instead of `this`. That seemed like it was at least avoiding throwing exceptions, but I didn't test without my CF change to see if it rendered the CF change unnecessary. (I had been trying to fix a different problem, and my `createFlowAnalysis` hack didn't help with that.))

I don't know if anything could go wrong as a result of such a change, but I wouldn't be shocked if it violated some assumption somewhere. Then again, maybe even my change to CF itself can violate some assumption somewhere :)
